src/System.Text.Json/src/System/Text/Json/JsonDocument.Parse.cs

```diff
@@ -0,0 +1,280 @@
+
+            CheckSupportedOptions(readerOptions);
+
+            return ParseAsyncCore(utf8Json, readerOptions, cancellationToken);
+        }
+
+        private static async Task<JsonDocument> ParseAsyncCore(
+            Stream utf8Json,
+            JsonReaderOptions readerOptions = default,
+            CancellationToken cancellationToken = default)
+        {
+            ArraySegment<byte> drained = await ReadToEndAsync(utf8Json, cancellationToken).ConfigureAwait(false);
+
+            try
+            {
+                return Parse(drained.AsMemory(), readerOptions, drained.Array);
+            }
+            catch
+            {
+                // Holds document content, clear it before returning it.
+                drained.AsSpan().Clear();
+                ArrayPool<byte>.Shared.Return(drained.Array);
+                throw;
+            }
+        }
+
+        public static JsonDocument Parse(ReadOnlyMemory<char> json, JsonReaderOptions readerOptions = default)
+        {
+            CheckSupportedOptions(readerOptions);
+
+            ReadOnlySpan<char> jsonChars = json.Span;
+            int byteCount = Utf8JsonReader.Utf8Encoding.GetByteCount(jsonChars);
+            byte[] utf8Bytes = ArrayPool<byte>.Shared.Rent(byteCount);
+
+            try
+            {
+                int byteCount2 = Utf8JsonReader.Utf8Encoding.GetBytes(jsonChars, utf8Bytes);
+                Debug.Assert(byteCount == byteCount2);
+
+                return Parse(utf8Bytes.AsMemory(0, byteCount2), readerOptions, utf8Bytes);
+            }
+            catch
+            {
+                // Holds document content, clear it before returning it.
+                utf8Bytes.AsSpan(0, byteCount).Clear();
+                ArrayPool<byte>.Shared.Return(utf8Bytes);
+                throw;
+            }
+        }
+
+        public static JsonDocument Parse(string json, JsonReaderOptions readerOptions = default)
+        {
+            CheckSupportedOptions(readerOptions);
+
+            return Parse(json.AsMemory(), readerOptions);
+        }
+
+        private static JsonDocument Parse(
+            ReadOnlyMemory<byte> utf8Json,
+            JsonReaderOptions readerOptions,
+            byte[] extraRentedBytes)
+        {
+            ReadOnlySpan<byte> utf8JsonSpan = utf8Json.Span;
+            Utf8JsonReader reader = new Utf8JsonReader(
+                utf8JsonSpan,
+                isFinalBlock: true,
+                new JsonReaderState(JsonReaderState.DefaultMaxDepth, readerOptions));
+
+            var database = new CustomDb(utf8Json.Length);
+            var stack = new StackRowStack(JsonReaderState.DefaultMaxDepth * StackRow.Size);
+
+            try
+            {
+                Parse(utf8JsonSpan, reader, ref database, ref stack);
+
+                if (database.Length == 0)
+                {
+                    throw new JsonReaderException("Cannot load the empty document", -1, -1);
```

Comment:
Move to resx?

Reply:
This is only needed until the PR for #34155 is merged.  I've added a TODO comment for removing this if altogether since that change seems fairly imminent.